chore: Add NetworkObject ownership docs and update related docs (#3879)
MidLevel/MLAPI: com.unity.netcode.gameobjects/Documentation~/advanced-topics/networkobject-ownership.md

# NetworkObject ownership

> Before reading these docs, familiarize yourself with the concepts of [ownership](../terms-concepts/ownership.md) and [authority](../terms-concepts/authority.md) within Netcode for GameObjects. It's also recommended to read the documentation on the [NetworkObject](../components/core/networkobject.md).

Reviewer: ```suggestion
Manage NetworkObject ownership across different network topologies.

Before reading this documentation, you should review the concepts of [ownership](../terms-concepts/ownership.md) and [authority](../terms-concepts/authority.md) in Netcode for GameObjects. It's also recommended to read the general [NetworkObject documentation](../components/core/networkobject.md).

## Identify ownership
```
Our content model requires a single sentence at the start of a page, starting with a verb (it makes more sense in the new docs platform)

To see if the local client is the owner of a NetworkObject, you can check the[`NetworkObject.IsOwner`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.IsOwner.html) or the [`NetworkBehaviour.IsOwner`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkBehaviour.IsOwner.html) property.

> To see if the server owns a NetworkObject, you can check the [`NetworkObject.IsOwnedByServer`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.IsOwnedByServer.html) or the [`NetworkBehaviour.IsOwnedByServer`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkBehaviour.IsOwnedByServer.html) property.

Reviewer: ```suggestion
To identify whether the local client is the owner of a NetworkObject, you can check the[`NetworkObject.IsOwner`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.IsOwner.html) or the [`NetworkBehaviour.IsOwner`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkBehaviour.IsOwner.html) property.

To identify whether the server owns a NetworkObject, you can check the [`NetworkObject.IsOwnedByServer`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.IsOwnedByServer.html) or the [`NetworkBehaviour.IsOwnedByServer`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkBehaviour.IsOwnedByServer.html) property.
```
We shouldn't use 'see' here, for [accessibility reasons](https://docs-style-guide.unity.com/style/inclusive-language/accessible-language/ability-based-language)

> [!NOTE]

> > When you want a specific NetworkObject to keep existing after the owner leaves the session, you can set the `NetworkObject.DontDestroyWithOwner` property to `true`. This ensures that the owned NetworkObject isn't destroyed as the owner leaves.

Reviewer 2: Here are some suggested adjustments:
To assure a spawned NetworkObject persists after the owner leaves a session, set the `NetworkObject.DontDestroyWithOwner` property to `true`. This assures the client-owned NetworkObject doesn't get destroyed when the owning client leaves.

## Authority ownership

> If you're creating a client-server game and you want a client to control more than one NetworkObject, or if you're creating a distributed authority game and the authority/current owner of the object would like to change ownership, use the following ownership methods.

Reviewer 2: ## Ownership vs authority

The client-server and distributed authority network topologies have slight differences when it comes to ownership and authority.

- Client-server: A client can own a spawned NetworkObject but the server always maintains spawn authority over the NetworkObject. Ownership only provides the owning client authority over NetworkVariables (_with write permissions_), any NetworkTransform configured for owner authority, and can be used to send an RPC that targets the owner.
- Distributed authority: If a client owns a NetworkObject it becomes the "over-all" authority of that NetworkObject. This means the owning client has spawn/de-spawn authority, write permissions to NetworkVariables (_associated with the owned NetworkObject_), the motion authority (if the NetworkObject has any NetworkTransforms), and the owning client can change the ownership permissions flags.

The following methods change their behavior based on the network topology being used:

Reviewer: This, except one day Noel will learn that it's 'overall' not 'over-all'

> The default `NetworkObject.Spawn` method will set server-side ownership when using a client-server topology. When using a distributed authority topology, this method will set the client who calls the method as the owner.

Reviewer: ```suggestion
The default `NetworkObject.Spawn` method sets server-side ownership in a client-server topology. When using a distributed authority topology, this method sets the client who calls the method as the owner.
```

```csharp
GetComponent<NetworkObject>().Spawn();

> ```

Reviewer 2: ### NetworkObject.Spawn

- Client-Server: The server defaults as the owner of this spawned object.
- Distributed authority: The client invoking this method becomes the owner of the spawned object.

To change ownership, the authority uses the `ChangeOwnership` method:

```csharp

> GetComponent<NetworkObject>().ChangeOwnership(clientId);

Reviewer 2: NetworkObject.ChangeOwnership(clientId);

```

> 

Reviewer 2: ### NetworkObject.SpawnWithOwnership

- Client-Server: The client identifier passed in as a parameter defines who the owner will be.
- Distributed authority: It is not recommended to use this method when using distributed authority if there are initial settings you want applied when spawning the new instance. For this scenario, it is recommended to use `NetworkObject.Spawn`, allow the settings to be applied locally, and then use `NetworkObject.ChangeOwnership`.

Author: I moved SpawnWithOwnership to the bottom of the page. That way ChangeOwnership has a higher prominence than SpawnWithOwnership, hopefully leading to fewer SpawnAuthority issues.

## Client-server ownership

In a client-server game, to give ownership back to the server use the `RemoveOwnership` method:

```csharp
GetComponent<NetworkObject>().RemoveOwnership();
```

> [!NOTE]
> `RemoveOwnership` isn't supported when using a [distributed authority network topology](../../terms-concepts/distributed-authority.md).

## Distributed authority ownership

> When building a distributed authority game, it is important to remember that owner of an object is always the authority for that object. As the simulation is shared between clients, it is important that ownership can be easily passed between game clients. This enables:

Reviewer: ```suggestion
In a distributed authority topology, the owner of a NetworkObject is always the authority for that NetworkObject. Because the simulation is shared between clients, it's important that ownership can be easily passed between game clients. This enables:
```

1. Evenly sharing the load of the game simulation via redistributing objects whenever a player joins or leaves.
2. Allowing ownership to be transferred immediately to a client in situations where a client is interacting with that object and so wants to remove lag from those interactions.
3. Controlling when and object is safe and/or valid to be transferred.

> The authority of any object can always change ownership as outlined in [authority ownership](#authority-ownership).

Reviewer: ```suggestion
- Evenly sharing the load of the game simulation by redistributing NetworkObjects whenever a player joins or leaves.
- Allowing ownership to be transferred immediately to a client in situations where the client is interacting with that NetworkObject to remove lag from those interactions.
- Controlling when a NetworkObject is safe or valid to be transferred.

The authority of any NetworkObject can always change ownership, as outlined in [authority ownership](#authority-ownership).
```
Numbered steps only for step by step processes. Bullet points for general list items.

> ### Ownership permissions settings

Reviewer: ```suggestion
### Ownership permission settings
```

> The following ownership permission settings, defined by [`NetworkObject.OwnershipStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipStatus.html), control how ownership of objects can be changed during a distributed authority session:

Reviewer: ```suggestion
The following ownership permission settings, defined by [`NetworkObject.OwnershipStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipStatus.html), control how ownership of NetworkObjects can be changed during a distributed authority session:
```

| **Ownership setting** | **Description** |
|-------------------|-------------|
| `None` | Ownership of this NetworkObject can't be redistributed, requested, or transferred (a Player might have this, for example). |
| `Distributable` | Ownership of this NetworkObject is automatically redistributed when a client joins or leaves, as long as ownership is not locked or a request is pending. |
| `Transferable` | Any client can change ownership of this NetworkObject at any time, as long as ownership is not locked or a request is pending. |
| `RequestRequired` | Ownership of this NetworkObject must be requested before ownership can be changed. |
| `SessionOwner` | This NetworkObject is always owned by the [session owner](distributed-authority.md#session-ownership) and can't be transferred or distributed. If the session owner changes, this NetworkObject is automatically transferred to the new session owner. |

You can also use `NetworkObject.SetOwnershipLock` to lock and unlock the permission settings of a NetworkObject for a period of time, preventing ownership changes on a temporary basis.

```csharp
// To lock an object from any ownership changes

> GetComponent<NetworkObject>().SetOwnershipLock(true);

Reviewer 2: NetworkObject.SetOwnershipLock(true);

// To unlock an object so that the underlying ownership permissions apply again

> GetComponent<NetworkObject>().SetOwnershipLock(false);

Reviewer 2: NetworkObject.SetOwnershipLock(false);

```

### Changing ownership in distributed authority

When a NetworkObject is set with `OwnershipPermissions.Transferable` any client can change ownership to any other client using the `ChangeOwnership` method:

```csharp

> GetComponent<NetworkObject>().ChangeOwnership(clientId);

Reviewer 2: NetworkObject.ChangeOwnership(clientId);

// To change ownership to self

> GetComponent<NetworkObject>().ChangeOwnership(NetworkManager.Singleton.LocalClientId);

Reviewer 2: NetworkObject.ChangeOwnership(NetworkManager.LocalClientId);

```

When a non-authoritative game client calls `ChangeOwnership`, the ownership change can fail. On a failed attempt to change ownership, the `OnOwnershipPermissionsFailure` callback will be invoked with a [`NetworkObject.OwnershipPermissionsFailureStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipPermissionsFailureStatus.html) to give information on the failure.

```csharp
internal class MyTransferrableBehaviour : NetworkBehaviour
{
    public override void OnNetworkSpawn()
    {
        NetworkObject.OnOwnershipPermissionsFailure = OnOwnershipPermissionsFailure;
        base.OnNetworkSpawn();
    }

    private void OnOwnershipPermissionsFailure(NetworkObject.OwnershipPermissionsFailureStatus ownershipPermissionsFailureStatus)
    {
        // Called on the calling client when NetworkObject.ChangeOwnership() has failed
    }

    public override void OnNetworkDespawn()
    {
        NetworkObject.OnOwnershipPermissionsFailure = null;
        base.OnNetworkDespawn();
    }

> }

Reviewer 2: public class ChangeOwnershipBehaviour : NetworkBehaviour
{
    private Action OwnershipChanged;
    public override void OnNetworkSpawn()
    {
        NetworkObject.OnOwnershipPermissionsFailure += OnOwnershipPermissionsFailure;
        base.OnNetworkSpawn();
    }

    public void TakeOwnership(Action ownershipChanged)
    {
        if (OwnershipChanged != null)
        {
            Debug.LogWarning($"Ownership request already sent!");
            return;
        }
        OwnershipChanged = ownershipChanged;
        NetworkObject.ChangeOwnership(NetworkManager.LocalClientId);
    }

    protected override void OnOwnershipChanged(ulong previous, ulong current)
    {
        if (current == NetworkManager.LocalClientId)
        {
            // If this is invoked, then the ownership change succeeded
            try
            {
                OwnershipChanged?.Invoke();
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
            }
            OwnershipChanged = null;
        }
        base.OnOwnershipChanged(previous, current);
    }

    /// <summary>
    /// Invoked on the calling client when NetworkObject.ChangeOwnership() has failed
    /// </summary>
    /// <param name="ownershipPermissionsFailureStatus">Why the change in ownership failed.</param>
    private void OnOwnershipPermissionsFailure(NetworkObject.OwnershipPermissionsFailureStatus ownershipPermissionsFailureStatus)
    {
        OwnershipChanged = null;
        // Make any adjustments or provide player feedback based on why the change in ownership failed.
        switch (ownershipPermissionsFailureStatus)
        {
            case NetworkObject.OwnershipPermissionsFailureStatus.NotTransferrable:
                {
                    break;
                }
            case NetworkObject.OwnershipPermissionsFailureStatus.RequestInProgress:
                {
                    break;
                }
            case NetworkObject.OwnershipPermissionsFailureStatus.SessionOwnerOnly:
                {
                    break;
                }
            case NetworkObject.OwnershipPermissionsFailureStatus.RequestRequired:
                {
                    break;
                }
            case NetworkObject.OwnershipPermissionsFailureStatus.Locked:
                {
                    break;
                }
        }
    }

    public override void OnNetworkDespawn()
    {
        NetworkObject.OnOwnershipPermissionsFailure += OnOwnershipPermissionsFailure;
        base.OnNetworkDespawn();
    }
}

Author: Are we sure we want all this detailed code here? Maybe it'd be more useful to have the most simple example here and a more fully featured "putting it all together" section at the end?

```

### Requesting ownership in distributed authority

When a NetworkObject is set with `OwnershipPermissions.RequestRequired` any client can request the ownership for themselves using the `RequestOwnership` method:

```csharp

> var requestStatus = GetComponent<NetworkObject>().RequestOwnership();

Reviewer 2: var requestStatus = NetworkObject.RequestOwnership();

```

> The `RequestOwnership` will return an [`OwnershipRequestStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipRequestStatus.html) to indicate the initial status of the request. To view the result of the request, `OnOwnershipRequestResponse` callback will be invoked with a [`NetworkObject.OwnershipRequestResponseStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipRequestResponseStatus.html).

Reviewer: ```suggestion
`RequestOwnership` returns an [`OwnershipRequestStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipRequestStatus.html) to indicate the initial status of the request. To view the result of the request, the `OnOwnershipRequestResponse` callback will be invoked with a [`NetworkObject.OwnershipRequestResponseStatus`](https://docs.unity3d.com/Packages/com.unity.netcode.gameobjects@latest?subfolder=/api/Unity.Netcode.NetworkObject.OwnershipRequestResponseStatus.html).
```

```csharp
internal class MyRequestableBehaviour : NetworkBehaviour
{
    public override void OnNetworkSpawn()
    {
        NetworkObject.OnOwnershipRequestResponse = OnOwnershipRequestResponse;
        base.OnNetworkSpawn();
    }

    private void OnOwnershipRequestResponse(NetworkObject.OwnershipRequestResponseStatus ownershipRequestResponseStatus)
    {
        // Called when the requesting client has gotten a response to their request
    }

    public override void OnNetworkDespawn()
    {
        NetworkObject.OnOwnershipRequestResponse = null;
        base.OnNetworkDespawn();
    }

> }

Reviewer 2: /// <summary>
/// When a NetworkObject's permissions includes the
/// <see cref="NetworkObject.OwnershipStatus.RequestRequired"/> flag.
/// </summary>
public class RequestableOwnershipBehaviour : NetworkBehaviour
{
    public override void OnNetworkSpawn()
    {
        NetworkObject.OnOwnershipRequestResponse += OnOwnershipRequestResponse;
        base.OnNetworkSpawn();
    }

    private void OnOwnershipRequestResponse(NetworkObject.OwnershipRequestResponseStatus ownershipRequestResponseStatus)
    {
        switch (ownershipRequestResponseStatus)
        {
            case NetworkObject.OwnershipRequestResponseStatus.Approved:
                {
                    // OnOwnershipChanged and OnGainedOwnership is invoked on the client requesting side.
                    // OnOwnershipChanged and OnLostOwnership is invoked on the previous owning client side.
                    break;
                }
            case NetworkObject.OwnershipRequestResponseStatus.RequestInProgress:
                {
                    // Another client already has a request pending.
                    break;
                }
            case NetworkObject.OwnershipRequestResponseStatus.CannotRequest:
                {
                    // Cannot request means RequestRequired is no longer a permission of this NetworkObject.
                    // The owning client changed while the request was being sent.
                    break;
                }
            case NetworkObject.OwnershipRequestResponseStatus.Denied:
                {
                    // The owning client denied the request.
                    break;
                }
            case NetworkObject.OwnershipRequestResponseStatus.Locked:
                {
                    // Ownership has been locked by the owning client.
                    break;
                }
        }
    }

    public override void OnNetworkDespawn()
    {
        NetworkObject.OnOwnershipRequestResponse -= OnOwnershipRequestResponse;
        base.OnNetworkDespawn();
    }
}

```

## Spawn with ownership

> To spawn a `NetworkObject` that is [owned](../../terms-concepts/ownership.md) by a different game client than the one doing the spawning, use the following:

Reviewer: ```suggestion
To spawn a NetworkObject that's [owned](../../terms-concepts/ownership.md) by a different game client than the one doing the spawning, use the following:
```

```csharp

> GetComponent<NetworkObject>().SpawnWithOwnership(clientId);

Reviewer 2: NetworkObject.SpawnWithOwnership(clientId);

```

> [!NOTE]

> > Using `SpawnWithOwnership` can result in unexpected behaviour when the spawning game client makes any other changes on the object immediately after spawning.

Reviewer: ```suggestion
> Using `SpawnWithOwnership` can result in unexpected behavior when the spawning game client makes any other changes on the object immediately after spawning.
```
American English required 🥲

> Using `SpawnWithOwnership` and then editing the object locally will mean the client doing the spawning will behave as the "spawn authority". The spawn authority will have limited local [authority](../terms-concepts/authority.md) over the object, but will not have [ownership](../terms-concepts/ownership.md) of the object that is spawned. This means any owner-specific checks during the spawn sequence will not be invoked on the spawn authority side.

Reviewer: ```suggestion
Using `SpawnWithOwnership` and then editing the NetworkObject locally means that the client doing the spawning will behave as the spawn authority. The spawn authority has limited local [authority](../terms-concepts/authority.md) over the NetworkObject, but not [ownership](../terms-concepts/ownership.md) of the NetworkObject that's spawned. This means any owner-specific checks during the spawn sequence will not be invoked on the spawn authority side.
```

> Any time you would like to spawn an object for another client and then immediately make adjustments on that object, it's instead recommended to use `Spawn`. After adjusting, the spawn authority can immediately follow with a call to `ChangeOwnership`.

Reviewer: ```suggestion
If you want to spawn a NetworkObject for another client and then immediately make adjustments to that NetworkObject, it's recommended to use the `Spawn` method. After adjusting, the spawn authority can immediately follow with a call to `ChangeOwnership`.
```

```csharp
[Rpc(SendTo.Authority)]
void SpawnForRequestingPlayer(RpcParams rpcParams = default) {
  var instance = Instantiate(myprefab).GetComponent<NetworkObject>();
  instance.Spawn();
  instance.transform.position = transform.position;
  instance.GetComponent<MyNetworkBehaviour>.MyNetworkVar.Value = initialValue;
  instance.ChangeOwnership(rpcParams.Receive.SenderClientId)
}
```

> This flow allows the spawning client to completely spawn and finish initializing the object locally before transferring the ownership to another game client.

Reviewer: ```suggestion
This flow allows the spawning client to completely spawn and finish initializing the NetworkObject locally before transferring the ownership to another game client.

## Additional resources

- [NetworkObject](../components/core/networkobject.md)
- [Ownership](../terms-concepts/ownership.md)
- [Authority](../terms-concepts/authority.md)

```
